src/tribler/core/utilities/slow_coro_detection.py

```diff
@@ -0,0 +1,136 @@
+import io
+import logging
+import time
+from asyncio import Handle, Task
+from threading import Lock, Thread
+
+from typing import Optional
+
+from tribler.core.utilities.main_thread_stack_tracking import get_main_thread_stack
+
+from tribler.core.utilities.main_thread_stack_tracking import main_stack_tracking_is_activated
+
+# pylint: disable=protected-access
+
+logger = logging.getLogger(__name__)
+
+SLOW_CORO_DURATION_THRESHOLD = 1.0
+WATCHING_THREAD_INTERVAL = 1.0
+
+
+class DebugInfo:
+    def __init__(self):
+        self.handle = None
+        self.start_time = None
+
+
+current = DebugInfo()
+lock = Lock()
+
+
+class SlowCoroWatchingThread(Thread):
+    """
+    A thread that detects and reports slow coroutines.
+    """
+    def run(self):
+        prev_reported_handle = None  # to detect in the loop when we have the second report of the same slow coroutine
+        while True:
+            time.sleep(WATCHING_THREAD_INTERVAL)
+            with lock:
+                handle, start_time = current.handle, current.start_time
+            if handle is not None:
+                duration = time.time() - start_time
+                if duration > SLOW_CORO_DURATION_THRESHOLD:
+                    report_freeze(current.handle, duration, first_report=(prev_reported_handle is not handle))
+                    prev_reported_handle = handle
+                else:
+                    prev_reported_handle = None
+
+
+def patched_handle_run(self: Handle):
+    """
+    Remembers the current asyncio handle object and its starting time globally, so it becomes possible
+    to access it from the separate thread and detect slow coroutines.
+    """
+    start_time = time.time()
+    with lock:
+        current.handle, current.start_time = self, start_time
+    try:
+        original_handle_run(self)
+    finally:
+        with lock:
+            assert current.handle is self, "asyncio loop should never run one handle inside another one"
```

Comment:
Is this assert is really necessary? What is the purpose of it?

Reply:
Asserts are handy for declaring non-trivial invariants that should always be satisfied if the code does not contain a bug. They can serve both as executable comments "the code always satisfies this condition" and runtime checks that allows the detection of rare bugs if an invariant is not satisfied.

I removed it as it is not strictly necessary to have it.